core/lensHelpers.ml

```diff
@@ -0,0 +1,302 @@
+    | _ -> failwith ("Unknown lens (get_primary_key) : " ^ (string_of_value lens))
+
+let rec lens_get_mem (lens : Value.t) callfn =
+    match lens with
+    | `Lens _ -> failwith "Non memory lenses not implemented."
+    | `LensMem (table, _rtype) -> table
+    | `LensDrop (lens, drop, _key, _def, _rtype) ->
+        let records = lens_get_mem lens callfn in
+        let result = List.map (fun a -> drop_record_column drop a) (unbox_list records) in
+          `List result
+    | `LensSelect (lens, pred, _sort) ->
+        let records = lens_get_mem lens callfn in
+        let res = List.filter (fun x -> unbox_bool (calculate_predicate_rec pred x)) (unbox_list records) in 
+           box_list res
+    | `LensJoin (lens1, lens2, on, _left, _right, _sort) ->
+        let records1 = lens_get_mem lens1 callfn in
+        let records2 = lens_get_mem lens2 callfn in
+        let on = List.map (fun (a, _, _) -> a) on in
+        let output = List.map (fun r1 -> 
+            let rows = List.filter (fun r2 -> is_row_cols_record_match r1 r2 (ColSet.of_list on)) (unbox_list records2) in
+            List.map (fun r2 ->  join_records r1 r2 on) rows
+        ) (unbox_list records1) in
+        box_list (List.concat output)
+    | _ -> failwith "Not a lens."
+
+let rec lens_get_db (lens : Value.t) =
+    match lens with 
+    | `Lens (((db, _), _, _, _), _) -> db
+    | `LensDrop (l, _, _, _, _) -> lens_get_db l
+    | `LensSelect (lens, _, _) -> lens_get_db lens
+    | `LensJoin (l1, _, _, _, _, _) -> lens_get_db l1
+    | _ -> failwith "Unsupported lens for get db."
+
+let rec lens_get_query (lens : Value.t) =
+  match lens with
+  | `Lens (((db, _), table, _, _), sort) -> 
+        let cols = LensSort.cols sort in
+        {
+            tables = [table, table];
+            cols = cols;
+            pred = None;
+            db = db;
+        }
+  | `LensSelect (lens, pred, _sort) ->
+        let query = lens_get_query lens in
+        { query with pred = Some pred }
+        (* get_lens_sort_row_type sort *)
+  | `LensJoin (lens1, lens2, _on, _left, _right, sort) ->
+        let q1 = lens_get_query lens1 in
+        let q2 = lens_get_query lens2 in
+        (* all table names must be unique, rename them *)
+        let _tables2 = List.map (fun (n2, al2) -> 
+            try 
+                let _tbl = List.find (fun (n1,_al1) -> n1 = n2) q1.tables in
+                failwith "Cannot reuse a table twice in a join query!"
+            with
+                NotFound _ -> (n2, al2)
+        ) q2.tables in
+        let tables = List.append q1.tables q2.tables in
+        let cols = LensSort.cols sort in
+        if (q1.db <> q2.db) then
+            failwith "Only single database expressions supported."
+        else
+            {tables = tables; cols = cols; pred = get_lens_sort_pred sort; db = q1.db}
+  | _ -> failwith "Unsupported lens for query"
+
+(* BUG: Lists can be too big for List.map; need to be careful about recursion *)
```

Comment:
It looks like you are mapping over a (potentially large) list of, say, length `N`. The `List.map` function uses `O(N)` stack space `O(N)`. So you indeed risk overflowing the stack for large lists. 

You can implement a space efficient `map` that uses `O(1)` stack space by using a "double fold". That is first transform the data using a `List.fold_left`, the elements will be reversed in the resulting list. You rectify this by using another `List.fold_left` on the result list.

What is the complexity of this? The time complexity is clearly `O(2N) = O(N)` (assuming that the transformation function is well-behaved and does not capture the list that we are mapping over). Since OCaml is a strict language, you have the property that `List.fold_left` is tail-recursive, allowing tail call optimisation to kick in, which means that `List.fold_left` uses `O(1)` space. 

I believe that the `List.rev_map` function is implemented in terms of `List.fold_left`, thus you ought to be able to implement the map described above as simple as

```ocaml
let space_efficient_map f xs = 
  List.(rev_map (fun x -> x) (rev_map f xs))
```

If you are willing to use side-effects and you know `N` in advance, then you can obviously do it in one linear pass using a for loop and mutation.

Reply:
If the lists here are being used/generated internally to represent sets of tuples (whose order doesn't matter) then just using rev_map would also be fine.